tsdb/index/metric_uid.go

```diff
@@ -0,0 +1,195 @@
+package index
+
+import (
+	"encoding/binary"
+	"sync"
+
+	"go.uber.org/zap"
+
+	"github.com/eleme/lindb/kv"
+	"github.com/eleme/lindb/pkg/constants"
+	"github.com/eleme/lindb/pkg/logger"
+	"github.com/eleme/lindb/pkg/util"
+	"github.com/eleme/lindb/tsdb/tree"
+)
+
+//MetricUID represents metric globally unique under a database
+type MetricUID struct {
+	metrics    map[uint32]*tree.BTree //Key is the ascii of the first letter of measurement
+	sequenceID uint32                 //unique sequence id
+	family     kv.Family
+	rwLock     sync.RWMutex //rwMutex of MetricUID
+}
+
+//NewMetricUID creation requires kvFamily
+func NewMetricUID(kvFamily kv.Family) *MetricUID {
+	//init sequenceID
+	id := getLastSequenceID(kvFamily)
+	mu := &MetricUID{
+		metrics:    make(map[uint32]*tree.BTree),
+		sequenceID: id,
+		family:     kvFamily,
+	}
+
+	return mu
+}
+
+//Finds the ID associated with a given name or creates it.
+func (m *MetricUID) GetOrCreateMetricID(metricName string, create bool) (uint32, bool) {
+	if len(metricName) > 0 {
+		nameBytes := []byte(metricName)
+		partition := getPartition(nameBytes)
+
+		id := m.getMeasurementIDFromDisk(partition, nameBytes)
+		if id == constants.NotFoundMetricID {
+			// if not exists
+			measurementTree, ok := m.metrics[partition]
+			if !ok {
+				//
+				measurementTree = tree.NewBTree()
+				m.metrics[partition] = measurementTree
+			}
+			if create {
+				m.rwLock.Lock()
+				defer m.rwLock.Unlock()
+
+				m.sequenceID++
+				measurementTree.Put(nameBytes, int(m.sequenceID))
+				return m.sequenceID, true
+			}
+			return constants.NotFoundMetricID, false
+		}
+		return id, true
+	}
+	return constants.NotFoundMetricID, false
+}
+
+func (m *MetricUID) SuggestMetrics(prefix string, limit int16) map[string]struct{} {
+	if len(prefix) > 0 {
+		nameBytes := []byte(prefix)
+		partition := getPartition(nameBytes)
+		snapshot, err := m.family.GetSnapshot(partition)
+		if nil == err {
```

Comment:
Nesting level too deep!
how about use this
`
if err!= nil {
return nil
}
if readers == nil{
return nil
}
`